File: apps/local-npm-mcp-server/src/gateway.ts
// Authenticated HTTP client to the API Gateway (the single source of truth). The user's
// `pk_live_` key is sent as a Bearer token so scoped writes (book_meeting) are authorized.
import { config } from "./config";

export class GatewayError extends Error {
  status?: number;
  constructor(message: string, status?: number) {
    super(message);
    this.name = "GatewayError";
    this.status = status;
  }
}

export const hasApiKey = (): boolean => Boolean(config.apiKey);

function buildHeaders(): Record<string, string> {
  const headers: Record<string, string> = {
    "Content-Type": "application/json",
    Accept: "application/json",
  };
  if (config.apiKey) headers.Authorization = `Bearer ${config.apiKey}`;
  return headers;
}

async function request(path: string, init: RequestInit, timeoutMs: number): Promise<unknown> {
  const controller = new AbortController();
  const timer = setTimeout(() => controller.abort(), timeoutMs);
  try {
    const res = await fetch(`${config.apiBase}${path}`, {
      ...init,
      headers: buildHeaders(),
      signal: controller.signal,
    });

    const raw = await res.text();
    let body: unknown;
    try {
      body = raw ? JSON.parse(raw) : undefined;
    } catch {
      body = raw;
    }

    if (!res.ok) {
      let message = `Gateway responded ${res.status}`;
      if (body && typeof body === "object" && "message" in body) {
        message = String((body as { message: unknown }).message);
      }
      throw new GatewayError(message, res.status);
    }
    return body;
  } catch (err) {
    if (err instanceof GatewayError) throw err;
    if (err instanceof Error && err.name === "AbortError") {
      throw new GatewayError(`Request to the gateway timed out after ${timeoutMs}ms`);
    }
    throw new GatewayError(`Network error reaching the gateway: ${(err as Error).message}`);
  } finally {
    clearTimeout(timer);
  }
}

export async function gatewayGet(
  path: string,
  params?: Record<string, string | undefined>
): Promise<unknown> {
  let qs = "";
  if (params) {
    const entries = Object.entries(params).filter(([, v]) => v != null) as [string, string][];
    if (entries.length) qs = "?" + new URLSearchParams(entries).toString();
  }
  return request(path + qs, { method: "GET" }, 15000);
}

export async function gatewayPost(path: string, body: unknown): Promise<unknown> {
  return request(path, { method: "POST", body: JSON.stringify(body) }, 25000);
}

/** Pull `.data` out of the gateway's `{ data: ... }` envelope, with a fallback. */
export function unwrap<T>(res: unknown, fallback: T): T {
  if (res && typeof res === "object" && "data" in res) {
    return (res as { data: T }).data ?? fallback;
  }
  return fallback;
}

File: apps/local-python-mcp-server/src/prajwalraj_mcp/gateway.py
"""Authenticated HTTP client to the API Gateway (single source of truth). The user's
`pk_live_` key is sent as a Bearer token so scoped writes (book_meeting) are authorized.
Mirrors apps/ai-service/src/tools/gateway.py.
"""
from typing import Any

import httpx

from . import config


class GatewayError(Exception):
    def __init__(self, message: str, status: int | None = None):
        super().__init__(message)
        self.status = status


def _headers() -> dict[str, str]:
    headers = {"Content-Type": "application/json", "Accept": "application/json"}
    if config.API_KEY:
        headers["Authorization"] = f"Bearer {config.API_KEY}"
    return headers


async def _request(
    method: str,
    path: str,
    *,
    params: dict | None = None,
    json: dict | None = None,
    timeout: float,
) -> Any:
    try:
        async with httpx.AsyncClient(timeout=timeout) as client:
            resp = await client.request(
                method, f"{config.API_BASE}{path}", params=params, json=json, headers=_headers()
            )
    except httpx.HTTPError as exc:
        raise GatewayError(f"Network error reaching the gateway: {exc}") from exc

    if resp.status_code >= 400:
        message = f"Gateway responded {resp.status_code}"
        try:
            body = resp.json()
            if isinstance(body, dict) and body.get("message"):
                message = str(body["message"])
        except Exception:
            pass
        raise GatewayError(message, resp.status_code)

    try:
        return resp.json()
    except Exception:
        return {}


async def gateway_get(path: str, params: dict | None = None) -> Any:
    return await _request("GET", path, params=params, timeout=15.0)


async def gateway_post(path: str, json: dict) -> Any:
    return await _request("POST", path, json=json, timeout=25.0)


def unwrap(res: Any, fallback: Any) -> Any:
    """Pull `.data` out of the gateway's `{ data: ... }` envelope, with a fallback."""
    if isinstance(res, dict) and "data" in res:
        data = res.get("data")
        return fallback if data is None else data
    return fallback
